### PR TITLE
fix(a11y): datagrid detail pane has role dialog with no aria label

### PR DESCRIPTION
**Pull request by andyfeds**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
`aria-describedby` is used on the datagrid detail pane instead of `aria-labelledby` 

Issue Number: [CDE-645](https://jira.eng.vmware.com/browse/CDE-645)

## What is the new behavior?

- By default if the user does not add anything: `aria-labelledby` will be added by the library with the id of the modal title.
- The user can set additional `aria-labelledby` (using the [clrDetailAriaLabelledBy] input) which could have a value referencing other text on the page , it can have multiple space separated id's. In this case the default title id will be added along with the user defined IDs.
- The user can add `aria-label` (using [clrDetailAriaLabel] input), in this case the aria-label provided will be used. 
- If both `aria-label` & `aria-labelledby` is added by the user, then only `aria-labelledby` will be added to the modal.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/angular/src/data/datagrid/datagrid-detail.ts [hunk 1]
```diff
@@ -38,8 +39,27 @@ import { DetailService } from './providers/detail.service';
...
   @ContentChild(ClrDatagridDetailHeader) header: ClrDatagridDetailHeader;
 
+  @Input('clrDetailAriaLabelledBy') ariaLabelledBy: string;
+  @Input('clrDetailAriaLabel') ariaLabel: string;
+
   constructor(public detailService: DetailService, public commonStrings: ClrCommonStringsService) {}
 
+  get labelledBy(): string {
+    if (this.ariaLabelledBy) {
+      return this.header ? this.header.titleId + ' ' + this.ariaLabelledBy : this.ariaLabelledBy;
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/datagrid-detail.ts:49]:
```suggestion
      return this.header ? `${this.header.titleId} ${this.ariaLabelledBy}` : this.ariaLabelledBy;
```

File: projects/angular/src/data/datagrid/datagrid-detail.ts [hunk 2]
```diff
@@ -38,8 +39,27 @@ import { DetailService } from './providers/detail.service';
...
+    } else {
+      return this.ariaLabel ? null : this.header ? this.header.titleId : '';
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/datagrid-detail.ts:51]:
```suggestion
      return this.ariaLabel || this.header?.titleId || '';
```

This is much easier to read than a nested ternary expression.

Comment (kevinbuhmann) [reply projects/angular/src/data/datagrid/datagrid-detail.ts:51]: Also, `this.ariaLabel` is not valid for `aria-label` because it is not a list of element IDs.

File: projects/angular/src/data/datagrid/datagrid-detail.ts [hunk 3]
```diff
@@ -27,7 +27,8 @@ import { DetailService } from './providers/detail.service';
       role="dialog"
       [id]="detailService.id"
       aria-modal="true"
-      [attr.aria-describedby]="header ? header.titleId : ''"
+      [attr.aria-labelledby]="labelledBy"
+      [attr.aria-label]="label"
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/datagrid-detail.ts:31]: Only one of these should be set.

"Don't use both on the same element because aria-labelledby will take precedence over aria-label if both are applied."
 https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label#:~:text=Don%27t%20use%20both%20on,label%20if%20both%20are%20applied.

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]

File: projects/angular/src/data/datagrid/datagrid-detail.ts (hunk 3 above)
Comment (andyfeds) [reply projects/angular/src/data/datagrid/datagrid-detail.ts:31]: By default `aria-labelledby` is added by the component, We need to give the user the flexibility to add an `aria-label` incase they do not want to use `aria-labelledby`. So if the user adds an `aria-label` using the input [clrDetailAriaLabel]="myLabel", then we do not set `aria-labelledby` and set only `aria-label`.

File: projects/angular/src/data/datagrid/datagrid-detail.ts (hunk 2 above)
Comment (kevinbuhmann) [reply projects/angular/src/data/datagrid/datagrid-detail.ts]: Oh, I misread the code.

File: projects/angular/src/data/datagrid/datagrid-detail.ts (hunk 3 above)
Comment (kevinbuhmann) [reply projects/angular/src/data/datagrid/datagrid-detail.ts:31]: Okay, I didn't realize one of the attributes isn't applied even though it is bound.

File: projects/angular/src/data/datagrid/datagrid-detail.ts [hunk 4]
```diff
@@ -38,8 +39,32 @@ import { DetailService } from './providers/detail.service';
...
+    if (this.ariaLabelledBy) {
+      return this.header ? `${this.header.titleId} ${this.ariaLabelledBy}` : this.ariaLabelledBy;
+    } else {
+      // If aria-label is set by the end user, do not set aria-labelledby
+      if (this.ariaLabel) {
+        return null;
+      } else {
+        return this.header?.titleId || '';
+      }
+    }
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/datagrid-detail.ts:57]:
```suggestion
    } else if (this.ariaLabel) {
      // If aria-label is set by the end user, do not set aria-labelledby.
      return null;
    } else {
      return this.header?.titleId || '';
    }
```

This can be flattened.

File: projects/angular/src/data/datagrid/datagrid-detail.ts [hunk 5]
```diff
@@ -38,8 +39,32 @@ import { DetailService } from './providers/detail.service';
...
+  }
+
+  get label(): string {
+    if (!this.ariaLabelledBy) {
+      return this.ariaLabel ? this.ariaLabel : null;
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/datagrid-detail.ts:62]:
```suggestion
      return this.ariaLabel || null;
```

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]

File: projects/angular/src/data/datagrid/datagrid-detail.ts (hunk 4 above)
Comment (andyfeds) [reply projects/angular/src/data/datagrid/datagrid-detail.ts]: Changed as suggested.

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]
Please add unit tests including tests to ensure both attributes are never both present and that `aria-labelledby` takes precedence over `aria-label`.

**Comment by andyfeds** [comment]
> Please add unit tests including tests to ensure both attributes are never both present and that `aria-labelledby` takes precedence over `aria-label`.

Added unit tests

File: projects/angular/src/data/datagrid/all.spec.ts [hunk 6]
```diff
@@ -81,6 +82,7 @@ describe('Datagrid', function () {
     DatagridCellSpecs();
     DatagridFilterSpecs();
     DatagridColumnSpecs();
+    DatagridDetailSpecs();
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/all.spec.ts:85]: I think the tests should be included and fixed separately to avoid including unrelated test fixes in the same commit as this bug fix. That can be confusing to readers. I was confused at first. I have opened #1451 for this.

File: projects/angular/src/data/datagrid/datagrid-detail.spec.ts [hunk 7]
```diff
@@ -53,25 +54,25 @@ export default function (): void {
         expect(context.clarityElement.innerHTML).not.toContain(content);
       });
 
-      it('hides content with the esc key', async(() => {
+      it('hides content with the esc key', waitForAsync(() => {
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/datagrid-detail.spec.ts:57]:
```suggestion
      it('hides content with the esc key', async(() => {
```

This is another unrelated change. Please revert. It will be fixed when #1440 is merged.

File: projects/angular/src/data/datagrid/all.spec.ts (hunk 6 above)
Comment (andyfeds) [reply projects/angular/src/data/datagrid/all.spec.ts:85]: Oh yes, I meant to discuss this with you. The older tests were commented out and were failing.
Makes sense to move it to a new PR. I have approved https://github.com/vmware-clarity/ng-clarity/pull/1451
Please check.

File: projects/angular/src/data/datagrid/datagrid-detail.spec.ts (hunk 7 above)
Comment (andyfeds) [reply projects/angular/src/data/datagrid/datagrid-detail.spec.ts]: Done

File: projects/angular/src/data/datagrid/all.spec.ts (hunk 6 above)
Comment (kevinbuhmann) [reply projects/angular/src/data/datagrid/all.spec.ts:85]: Thanks, my PR is merged now.

File: projects/angular/src/data/datagrid/datagrid-detail.spec.ts [hunk 8]
```diff
@@ -89,3 +128,16 @@ export default function (): void {
...
+    <clr-dg-detail [clrDetailAriaLabelledBy]="ariaLabelledBy" [clrDetailAriaLabel]="ariaLabel">
+      <clr-dg-detail-header>Title</clr-dg-detail-header>
+      <clr-dg-detail-body>${content}</clr-dg-detail-body>
+    </clr-dg-detail>
+  `,
+})
+class DataGridPaneTestComponent {
+  ariaLabelledBy: string;
+  ariaLabel: string;
+}
```
Comment (kevinbuhmann) [inline projects/angular/src/data/datagrid/datagrid-detail.spec.ts:143]: I think you can just merge this into `FullTest`.

**Review by kevinbuhmann — APPROVED** [review]
